### PR TITLE
Implement /etc/locale.conf support

### PR DESCRIPTION
**Pull request by davispuh**
When `wsetlocale(LC_ALL, L"")` is used then libc will use `$LANG` to get locale and if it's not set then it defaults to `C` locale which will make some C functions to fail when they encounter non-ASCII strings.

It can't be set later or in `etc/config.fish` because by then fish will already have used C functions which expect `$LANG`.

This fixes several locale/encoding related issues like #277, #1749 

`/etc/locale.conf` is correct file for system-wide locale settings, when systemd is used then `localectl set-locale` will simply write to that file. See more [man locale.conf](http://www.freedesktop.org/software/systemd/man/locale.conf.html)
`bash` also gets locale from there.

`su -l` keeps only `HOME, SHELL, USER, LOGNAME, PATH` and doesn't keep `LANG` so it won't be set for fish and that's why need to set it ourselves.

### CODE REVIEW

**Comment by zanchey** [comment]
My understanding is that it is the job of the login process - be that `ssh`, `su`, `login`, or your display manager - to set the locale of the processes it starts. For example, `su` is configured on Debian to read `/etc/default/locale`, `ssh` and `sudo` on OS X 

I don't agree that fish should read `/etc/locale.conf` - it's for systemd, and its contents can be overidden in systemd by kernel boot parameters.

Neither bash nor zsh read this file or attempt to find a locale that isn't in the environment, so I don't want to accept the reading of `/etc/locale.conf`. I would like to drop the locale stuff in `etc/config.fish`, though.

**Comment by davispuh** [comment]
Well, it would be nice if login process would do it, but it's not the case here on ArchLinux. Like I said it's shell's responsibility and it's always been so.

I disabled `/etc/profile` and used `bash`, here are all env vars when logged in with `su -l` (it's about same with `login` or `sudo`)

```
SHELL=/bin/bash
TERM=xterm-256color
USER=user
PATH=/usr/bin
PWD=/home/user
SHLVL=1
HOME=/home/user
LOGNAME=user
_=/usr/bin/env
```

and those are all, no LANG, thus `C` locale it is.

When logged in over `ssh`

```
XDG_SESSION_ID=c9
TERM=xterm-256color
SHELL=/bin/bash
SSH_CLIENT=::1 47982 22
SSH_TTY=/dev/pts/2
USER=user
MAIL=/var/spool/mail/user
PATH=/usr/bin:/bin:/usr/sbin:/sbin
PWD=/home/user
SHLVL=1
HOME=/home/user
LOGNAME=user
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1004/bus
SSH_CONNECTION=::1 47982 ::1 22
XDG_RUNTIME_DIR=/run/user/1004
_=/usr/bin/env
```

same, no `LANG` and this is why `fish` is unusable on my system without this PR.

The only place where I'm configuring locale is `/etc/locale.conf` and systemd's `localectl` writes to this file as it's system-wide settings and it's not only for systemd but for all applications to use. Also overriding it with kernel parameter is a feature, it's good way to test other locales without making permanent changes.

`bash` executes `/etc/profile` which pulls in `/etc/profile.d/*` and there is `locale.sh` which does

```bash
#!/bin/sh
if [ -z "$LANG" ]; then
  if [ -n "$XDG_CONFIG_HOME" ] && [ -r "$XDG_CONFIG_HOME/locale.conf" ]; then
    . "$XDG_CONFIG_HOME/locale.conf"
  elif [ -n "$HOME" ] && [ -r "$HOME/.config/locale.conf" ]; then
    . "$HOME/.config/locale.conf"
  elif [ -r /etc/locale.conf ]; then
    . /etc/locale.conf
  fi
fi
LANG=${LANG:-C}
export LANG
```

thus `bash` indirectly reads `/etc/locale.conf` but this is just a hack, because what if shell needs to use Unicode before it have read that file? it would fail and that's why it should be read at start before doing anything.

Also my display manager `sddm` does set locale but it needed same patch to do it sddm/sddm@80c3a87e154ac7d0b60c6a3a38b0ddee218b8e0b and it isn't used when I'm using fish in a tty or over ssh.

The main reason is that when `LANG` isn't found everything defaults to `C` locale which is just dumb in 2015, much better would be default to `UTF-8` but anyway this PR fixes it.

**Comment by zanchey** [comment]
I've asked the systemd-devel list about the stability of locale.conf. My feeling is that because there is such a wide range of methods of setting the locale that it probably needs to be sorted out by downstream distributions, but I appreciate that's not always straightforward. Work on #1956 would simplify things for our distribution maintainers.

**Comment by faho** [comment]
@zanchey: It's listed in the [Interface Portability And Stability Chart](http://www.freedesktop.org/wiki/Software/systemd/InterfacePortabilityAndStabilityChart/) as covered by the stability promise, fully documented and "Reimplementable Independently". So I think that's a yes.

I'm not sure how distributions that don't use systemd handle it - arch switched to those files before adopting systemd itself, but how about gentoo or slackware?

**Comment by davispuh** [comment]
Currently it's such a big mess because there's no standard for it and it's left up to user to set locale, but that doesn't really work... I think best is to just make `/etc/locale.conf`  de-facto standard as you don't really need systemd, but only that file and it will work out-of-box for systemd's users which are supported by large part of distributions like even Debian is migrating. Gentoo sets locale in ` /etc/env.d/02locale` but I'm not sure who sources that file.

#1956 would be useful indeed, but I think `/etc/fish` (or `fish.d`) would be preferable over `profile.d`. Anyway that still wouldn't solve this issue fully because fish's config is read too late when already Unicode might be needed as I mentioned previously. It would work fine for distributions and users who uses it later but not when it's required immediately after startup such as for file paths, env vars and so on. I'm not sure if it's achievable to source it very early because to parse fish need quite a lot of pre-setup.

**Comment by adrianratnapala** [comment]
I'm glad this bug merge is being considered.  The question of whether or not it is the shell's job to set LANG is a little irrelevant because in my installation of  versio 2.2.0 /etc/fish/config.fish checks it and sets it to en_US.UTF-8.

That is config.cfish already beleives it is job to set LANG, if it has not already been set.  It should certainly do something smarter than just forcing en_US.UTF-8.  Even C or POSIX would be better than that, since they at least exist on my system.

**Comment by zanchey** [comment]
Yes, that's definitely a bad idea.

**Comment by zanchey** [comment]
Unsurprisingly, there isn't really a right answer here.
 * systemd/systemd@1640944a84 describes what happens if you unconditionally assume the system locale is supported on the terminal
 * https://bugzilla.redhat.com/show_bug.cgi?id=663900#c12 points out that the system locale may be different to the user locale
 * http://lists.freedesktop.org/archives/systemd-devel/2015-October/034589.html explains that we can use `locale.conf`, but that there's no widely-established way of identifying the user locale - which is actually what we want to use

**Comment by faho** [comment]
>It can't be set later or in `etc/config.fish` because by then fish will already have used C functions which expect `$LANG`.

Is this actually still true? With fish 2.2.0-391-gbe0d9d0 (i.e. ToT) setting LANG to "C" interactively will complain about wide characters, but then running `fish` from that will not with `set -x LANG de_DE.UTF-8` in ~/.config/fish/config.fish, so it appears that setting LANG there is soon enough. I.e.:

```fish
> set LANG C
fish: Wide character 10148 has no narrow representation
# and a few more like it
# With set -x LANG de_DE.UTF-8 in config.fish:
> fish
> # No error
# Without:
> fish
fish: Tried to print invalid wide character string
```

Also, how would this fix #1749? That's about setting locale to a particular value (i.e. something that might not be the system-wide locale) in a particular source file.

**Comment by davispuh** [comment]
> https://bugzilla.redhat.com/show_bug.cgi?id=663900#c12 points out that the system locale may be different to the user locale

theoretically it can be different but practically that doesn't work at all. glibc and basically all applications can work only with 1 locale at a time (there's only one LANG env var)  but if you've different system's locale than user's locale then whenever application tries to access system's related functions/data it will use wrong "user locale" while it should use system's locale but no application separates it, there's just no such concept... So different user locale works only for those applications which don't interact with system or both locales uses same character encoding. Just think about is filesystem/paths/filenames a system locale or user locale?

> Is this actually still true?

yes, it doesn't work, because fish's config is parsed after reading env but to read env you already need to know what locale it uses and there's probably other functions which needs locale. Test with `su -l` to login which won't set LANG and ensure you've file paths containing Unicode.

for me, with fish just compiled from master I get this #2298 assert failure, but with #2321 merged I get

```
fish: env_set_pwd in env_init failed!
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
fish: Tried to print invalid wide character string
env
LANG=lv_LV.UTF-8
SHELL=/usr/bin/fish
fish: Tried to print invalid wide character string
```
as you can see I've correct locale set in `/etc/fish/config.fish` but it's getting set too late. Of course if I run `fish` again then it will work as it's set from parent fish before new fish have started.
Basically since September I'm using fish with this PR merged so that it atleast works for me.

> Also, how would this fix #1749? That's about setting locale to a particular value (i.e. something that might not be the system-wide locale) in a particular source file.

That user wanted to output login message in Unicode but because when fish is started on tty, LANG will be empty and config which he wrote in UTF-8 will be read as C locale making it wrong.
This fixes it by making so that fish's config will use `/etc/locale.conf` as locale and thus support reading Unicode config. Also I don't see any reason why anyone would want fish's config to be in different locale as it's in /etc anyway.

**Comment by krader1961** [comment]
@davispuh I respectfully disagree with pretty much everything you've written on this topic. It is perfectly reasonable for the user to set a locale different from whatever the sysadmin defined as the "system" locale and doing so should not cause any problems. Consider that there may be multiple user accounts and _user A_ wants messages in French and dates, times, etc. formatted in the usual French fashion while _user B_ wants messages in English but dates and times in ISO 8601 format rather than the usual english speaking conventions. In the distant past when EUC was in common use (extended unix code, something like MS Windows code pages) you definitely could see problems if your __LANG__ var was set to a character set different from the encoding used in various files (e.g., the sort command might produce unexpected ordering of the input).

Your examples involving filesystem path names should be irrelevant. If the system is storing the pathnames in a Unicode encoding it shouldn't matter what locale the shell is using. Modulo the obvious exception that the C locale will result in non-ASCII characters being treated as chars with the high bit set rather than Unicode.

You ask

> what if shell needs to use Unicode before it have read that file?

It shouldn't. There should no nothing but plain ASCII before the early initialization scripts (e.g., /etc/profile) are read.

Furthermore, the fish C++ code should not be reading distribution specific files like /etc/locale.conf. That is something that should be done in a config file that is sourced early in the initialization of the shell (e.g., your bash /etc/profile example). We might need to provide better defaults in the system level __config.fish__ file. Whether or not you consider setting env vars in a file like __config.fish__ "the shell's responsibility" is almost a philosophical question. I look at it as something that is the responsibility of the distro and the individual user (e.g., __$HOME/.config/fish/config.fish__). The shell is just doing what the distro maintainer or user told it to do. The shell itself isn't actually responsible for crafting an appropriate environment except in the most narrow sense of providing reasonable defaults so the shell will run if that customization hasn't been done by the distro or user.

As for your assertion that `su -l` removes __LANG__ from the environment that is only partially true. The __-l__ argument does (quoting from the Ubuntu su man page):

>The optional argument - may be used to provide an environment similar to what the user would expect had the user logged in directly.

The "optional argument -" is referring to the accepted alias for __-l__ for historical reasons. In other words __su__ when invoked with __-l__ discards the existing environment and should read distro specific files like /etc/locale.conf or /etc/default/local to setup an initial environment before launching the shell. If you don't use the hyphen option then it should leave the existing environment alone with the exception of __PATH__. That's what happens on every UNIX I've ever used. However, I've read enough issues involving unusual ArchLinux behavior I wouldn't be surprised if its maintainers have decided to radically deviate from historical practice in this regard.

**Comment by faho** [comment]
>Furthermore, the fish C++ code should not be reading distribution specific files like /etc/locale.conf

Not quite distribution-specific - it's a systemd thing, which means it's in Arch, Fedora, Debian, Ubuntu, OpenSUSE, .... Even if you're not using systemd, your distribution might still be using this particular file (I know that arch switched to using it before switching to systemd) - I'd have to check Slackware and Gentoo.

In other words: It's the best standard we currently have _on linux_.

>something that should be done in a config file that is sourced early in the initialization of the shell

I of course agree that it should be done in script, but the problem is that it appears to be needed earlier than we are reading config.fish now.

>However, I've read enough issues involving unusual ArchLinux behavior I wouldn't be surprised if its maintainers have decided to radically deviate from historical practice in this regard.

Arch uses util-linux's `su`, which says it should be clearing everything except TERM and initializing HOME, SHELL, USER, LOGNAME and PATH (see the [man page](http://www.man7.org/linux/man-pages/man1/su.1.html)). The rest is done by PAM, which does.... something.

When starting bash (`su -l -s /bin/bash`) I seem to be getting a $LANG from somewhere, and the only file I've set it in is /etc/locale.conf. When starting fish, I'm getting "en_US.UTF-8" - my LANG is actually "de_DE.UTF-8".

**Comment by krader1961** [comment]
While this change accommodates distros that use systemd, and a large number do so, that is still distro specific. Regardless, I don't see any good reason for this path to be baked into the core fish code. You say

> but the problem is that it appears to be needed earlier than we are reading config.fish now.

I cannot tell from the @davispuh's description of this change why setting __LANG__ or __LC_ALL__ needs to occur earlier than reading the global or personal __config.fish__ files. Have you reproduced the problem? I tried to given the ambiguous problem statement but was not able to do so. Which is why any non-trivial change should be tied to an open issue so this type of discussion can occur in the open issue. Comments on the pull-request should be about style, correctness, etc. rather than whether the change makes any sense.

The fact, @faho, that you're seeing __LANG__ set to __en_US.UTF-8__ is probably due to the dubious 

```
        if not set -q LANG >/dev/null
                set -gx LANG en_US.UTF-8
        end
```

in the fish global __config.fish__ file. It seems to me like that should explicitly set the LANG to C. Not to mention the fact it does not test for __LC_ALL__.

The distro specific bash initializations you noticed are probably setting __LANG__ to a value derived from __/etc/locale.conf__. Which is something I have no arguments against as long as they're doing as part of the scripted startup of the shell.

**Comment by faho** [comment]
> I cannot tell from the @davispuh's description of this change why setting LANG or LC_ALL needs to occur earlier than reading the global or personal config.fish files. Have you reproduced the problem?

Yes - with `set -gx LANG de_DE.UTF-8` in config.fish:

```fish
$ mkdir /tmp/ü
$ cd /tmp/ü
$ set -gx LANG C
$ fish
>fish: Could not determine current working directory. Is your locale set correctly?
# And then errors presumably resulting from that, including, after typing "c" (which presumably triggers a suggestion for /tmp/ü):
>fish: src/path.cpp:148: bool path_get_cdpath(const wcstring&, wcstring*, const wchar_t*, const env_vars_snapshot_t&): Assertion `wd[len - 1] == L'/'' failed.
>fish: 'fish' terminated by signal SIGABRT (Abort)
```

>The fact, @faho, that you're seeing LANG set to en_US.UTF-8 is probably due to the dubious 

Oh yeah, I missed that (only checked in share/), thanks!

>Not to mention the fact it does not test for LC_ALL.

Well, anything else should use LC_ALL if it is set - what we set LANG to shouldn't matter in that case.

**Comment by krader1961** [comment]
> Well, anything else should use LC_ALL if it is set - what we set LANG to shouldn't matter in that case.

It's not that simple. See http://pubs.opengroup.org/onlinepubs/7908799/xbd/envvar.html where it states that __LC_ALL__ supercedes __LANG__.  So if __LC_ALL__ is already set we should not need to set __LANG__; and while doing so in that case should be harmless it is confusing. This is one of those ugly details due to needing to retain backward compatibility. The question is whether a modern shell like fish should continue to use legacy mechanisms like __LANG__ or embrace their modern replacements like __LC_ALL__.

Regarding your recreation I can't replicate your results. With my PWD being the Unicode directory in your example the `set -gx LANG C` immediately resulted in multiple instances of this error:

```
fish: Wide character 776 has no narrow representation
```

Starting a new shell by executing `fish` reported no errors and `echo $LANG` in that shell reported `C`. The output after `cd ..` then `cd u[tab]` resulted in weird fish output (there were several forward-slash characters displayed) but no errors, per se, and the change directory into the unicode directory worked.

Clearly there are at least two problems (the one reported by @davispuh and the other by @faho). But this change does not appear to be the correct fix for either one. I think one or more issues should be opened that describe how to reproduce each problem.

**Comment by faho** [comment]
>LC_ALL supercedes LANG

In the sense that " If the LC_ALL environment variable is defined and is not null, the value of LC_ALL is used", yes. And that's how this can still work. But if a program doesn't read LC_ALL, I'm not sure it'll react nicely to a missing LANG.


What we do matches exactly what Arch's [locale.sh](https://projects.archlinux.org/svntogit/packages.git/tree/repos/core-x86_64/locale.sh?h=packages/filesystem) does. Not that arch is the end-all-be-all of shell programming, but their (bash) scripts are usually quite good. Anyway, this is a tangent.

>Regarding your recreation I can't replicate your results. With my PWD being the Unicode directory in your example the set -gx LANG C immediately resulted in multiple instances of this error:

Humm.... what fish version/OS is this on? I'm guessing a recent-ish build on OSX? I get this with empty configuration.

>fish: Wide character 776 has no narrow representation

I could have sworn we have a bug open about that. I've seen that rather often before. I'll open one.

>But this change does not appear to be the correct fix for either one.

That's probably right - though letting fish simply choose a locale without config for the simple case is a nice thing to have. That's also been asked in #277 - though that's asking for LC_ALL, which I'm against.

**Comment by faho** [comment]
After a bit more thinking on this, here's what I think would actually solve this issue, instead of papering over it in a specific case (linux with /etc/locale.conf), like this PR would do:

- Fix the crash in case of a non-ASCII PWD

- Disable the annoying "wide character has no narrow representation" messages at least on startup

Both of these combined would make setting LANG in config.fish work. locale.conf support could then look like

```fish
function readkeyvalue
	for file in $argv
		while read -l line
			set -l var (string split "=" -- $line)[1]
			set -l val (string split "=" -- $line)[2]
			set -gx $var $val
		end < $file
	end
end
test -r /etc/locale.conf; and readkeyvalue /etc/locale.conf
```

in etc/config.fish. This would make the simple case of single-locale systems with locale.conf (i.e. a large percentage of linux desktops) just work, and other similar files and schemes could trivially be supported - these could even be added locally by users.

I believe this PR to not really help very much. It'd be a stopgap.

**Comment by pickfire** [comment]
@krader1961 I am also ***against*** setting **LC_ALL** instead of **LANG**, not everyone set **LC_ALL**.

In https://wiki.archlinux.org/index.php/Locale#LC_ALL, setting **LC_ALL** will override **LANG** and all the other __LC_*__ variables, whether they are set or not. Not everyone want to have unified __LC_*__. For me I use `c` for **LC_MESSAGES** which is different from the others.

> set -gx LANG C

Why not use utf8 locale when you can?

**Comment by krader1961** [comment]
Upon further reflection I agree that fish should not set __LC_ALL__. I was under the misapprehension that vars like __LC_MESSAGES__ overrode __LC_ALL__. I did some more research and see that is not true.

> Why not use utf8 locale when you can?

Because there are still, sadly, environments that don't support Unicode we need to support. Consider a recently discussed issue involving the DEC VT-100/220/320/420 series of terminals used as the console of a Linux server. In the absence of an explicit declaration, by the sysadmin or user, that UTF-8 is supported fish should be conservative and work with plain ASCII.

**Comment by davispuh** [comment]
@krader1961 
> It is perfectly reasonable for the user to set a locale different from whatever the sysadmin defined as the "system" locale and doing so should not cause any problems. 
> [..]
> Your examples involving filesystem path names should be irrelevant. If the system is storing the pathnames in a Unicode encoding it shouldn't matter what locale the shell is using. 

It doesn't really work if encodings are different which I mentioned and it's huge assumption that they'll be in Unicode, if encodings are different then most files will show with wrong characters or as "????" and it would be complicated to enter such directories. Consider 2 users one with `LANG=ar_LY.ISO-8859-6` and other with `LANG=he_IL.ISO-8859-8`, there's nothing preventing users setting these, then system's locale is `en_US.UTF-8` (/etc/passwd uses UTF-8 and user's home directories are their native names). So before they login their home dirs will show up correctly but once they login they'll use wrong encoding because of user's LANG choice. This is reason why I say it can't work as it's not possible to separate system's and user's locale. Also when one user creates shared folders like "ناحي"  then for other user it will show wrong characters "ז??ך".
When I tested this with bash it worked ok and it was possible to enter such directories even when they're displayed wrong, but still I wouldn't call that really workable...
With fish (latest compiled from git master + this PR) it doesn't work at all when different locale is used, can't enter in those directories and fish even crashes (seems #2321 is actually required).

```
fish: Wide character 1589 has no narrow representation
fish: Wide character 1604 has no narrow representation
fish: Wide character 1613 has no narrow representation
fish: Could not determine current working directory. Is your locale set correctly?
cd: Could not set PWD variable
fish: Wide character 1608 has no narrow representation
fish: Wide character 1589 has no narrow representation
fish: Could not determine current working directory. Is your locale set correctly?
fish: Unable to create a configuration directory for fish. Your personal settings will not be saved. Please set the $XDG_CONFIG_HOME variable to a directory
where the current user has write access.
fish: Unable to create a configuration directory for fish. Your personal settings will not be saved. Please set the $XDG_CONFIG_HOME variable to a directory
where the current user has write access.
fish: Unable to create a data directory for fish. Your history will not be saved. Please set the $XDG_DATA_HOME variable to a directory where the current
user has write access.
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
fish: Could not determine current working directory. Is your locale set correctly?
fish: Unable to create a configuration directory for fish. Your personal settings will not be saved. Please set the $XDG_CONFIG_HOME variable to a directory
where the current user has write access.
fish: Unable to create a configuration directory for fish. Your personal settings will not be saved. Please set the $XDG_CONFIG_HOME variable to a directory
where the current user has write access.

test@arch > /fish: src/path.cpp:148: bool path_get_cdpath(const wcstring&, wcstring*, const wchar_t*, const env_vars_snapshot_t&): Assertion `wd[len - 1] == L'/'' failed.
fish: 'fish' terminated by signal SIGABRT (Abort)
``` 


> In other words su when invoked with -l discards the existing environment and should read distro specific files like /etc/locale.conf or /etc/default/local to setup an initial environment before launching the shell. 

That's not true, it's not `su` responsibility, it doesn't do that (see [util-linux/login-utils/su-common.c](http://git.kernel.org/cgit/utils/util-linux/util-linux.git/tree/login-utils/su-common.c#n971)), it simply discards environment and launches shell. `su` can't know what environment variables shell needs.


So anyway this PR tries to address only one thing. Currently `su -l` on ArchLinux clears environment and thus `LANG` is empty making it a shell's responsibility to set it. Current implementation just set's it as `en_US.UTF-8` and so this would be improvement, if there exists `/etc/locale.conf` it would use that LANG instead and if it doesn't then there's no difference.
There are several other locale related issues/bugs in fish but this PR isn't meant to address them, this is only about setting locale when there isn't one. And other issues still need to be solved so that fish would work even with wrong locale.

About why need LANG so early is because in [fish.cpp main()](/fish-shell/fish-shell/blob/master/src/fish.cpp#L505) calls `env_init` and it in [env.cpp env_init()](/fish-shell/fish-shell/blob/master/src/env.cpp#L537) calls `getpwnam` from which user's HOME directory is taken and it's in system's encoding so you need to know encoding to decode it.



> Furthermore, the fish C++ code should not be reading distribution specific files like /etc/locale.conf. That is something that should be done in a config file that is sourced early in the initialization of the shell (e.g., your bash /etc/profile example).

I don't agree with this, IMO this is best place where to read it, and I consider `/etc/locale.conf` as a standard for system's language location. Anyway this isn't my decision to make and I'm fine with what @faho proposed as long as it works correctly.

**Comment by krader1961** [comment]
The ISO 8859 standard is a nightmare. I was a UNIX support technician for [Sequent Computer Systems](https://en.wikipedia.org/wiki/Sequent_Computer_Systems) in the 1990's when it first started being widely deployed. It was a major source of pain for customers and vendor support staff. Anyone still using it is a masochist. There is absolutely no reason for fish to support it beyond what can be done trivially (such as by using the libc support for locales).

I wasn't as clear as I would have liked regarding my comments about file/path names. I was not referring to how they are displayed. If a system is supporting multiple users, each of which might be using a different single byte character encoding (e.g., a ISO 8859 encoding) there is no way to ensure every user sees the correct glyphs. This is why Unicode and encodings like UTF-8 were created. What fish, and other shells, should do is treat path names as binary byte sequences without a particular encoding. I believe this is what @faho is also recommending.

> I don't agree with this, IMO this is best place where to read it, and I consider /etc/locale.conf as a standard for system's language location.

What you personally consider a standard isn't relevant unless you can point to a real ISO, IEC, or other relevant standard. If you can do so and something like Mac OS X doesn't follow that standard then it's my responsibility to bitch at Apple Corp. Which is not the same thing as saying fish should not accommodate the quirks of as many mainstream UNIX flavors as possible. Only that we shouldn't hardcode into the core C++ code support for a particular flavor (e.g., ones using the relatively new systemd) unless there is absolutely no other option.

> su can't know what environment variables shell needs.

That is complete and utter nonsense. If the __su__ command didn't know what environment variables the program it launches needs why is it setting __TERM__, __HOME__, __SHELL__, etc. These things are simply conventions that a large group of people agree have meaning (which includes the maintainers of shells like bash and fish). The __LANG__ and __LC_*__ family of vars belong in the group of baseline environment vars that should be present for any shell. Programs like __su__ and __login__ are expected to follow those conventions.

**Comment by krader1961** [comment]
Also, read this http://unix.stackexchange.com/questions/39175/understanding-unix-file-name-encoding for a pretty good discussion of the issue regarding shells like fish handling path names.

**Comment by krader1961** [comment]
This pull-request is now more than five months old and hasn't had any comments or new commits in two months. It is also based on some incorrect premises and isn't tied to an open issue which is problematic.

@faho, @zanchey, What do you think since you've already commented on this PR? I vote that it be rejected in its present form. If people still believe fish should be doing this in either the core C++ code, or config scripts, then an enhancement issue should be opened where it can be discussed.

**Comment by faho** [comment]
>What do you think since you've already commented on this PR? I vote that it be rejected in its present form. If people still believe fish should be doing this in either the core C++ code, or config scripts, then an enhancement issue should be opened where it can be discussed.

I think if fish should have locale.conf support it should be added via script (e.g. as a snippet) and problems with not having the locale before then should be fixed. I don't think this PR is the right solution.

**Comment by zanchey** [comment]
My takeaway from the systemd list was that the shell probably does need to set the locale, but that it will need to be done on a distribution-specific level, so I agree that it should be done in shell initialisation files if at all possible.

**Comment by krader1961** [comment]
> My takeaway from the systemd list was that the shell probably does need to set the locale...

I don't understand why a system-wide default locale is any different from env vars like **PATH**, **HOME**, **USER**, etc. Such things are traditionally the responsibility of daemons like "init" (or, now, "systemd") and login managers like "xdm" and the PAM subsystem on Linux. My opinion of the systemd advocates has reached a new low.

**Comment by zanchey** [comment]
If you have a read of the list, Lennart acknowledges that this is a poor position. Unfortunately when systemd set the locale, some people were unable to log in, which made it something of a showstopper. The current state of affairs in other shells is that the locale is generally set by a distribution-specific initialisation script.

**Comment by davispuh** [comment]
All this time this PR has been needed for me and so I've been applying it locally.
I just tested without it and it's still needed. 

#3219 partially fixed it for most cases, but not for all because main problem is that locale setting happens way too late after `fish` already have read some environment variables/settings using POSIX locale and they aren't later reread after locale change.

For example when I SSH into my Arch Linux with latest fish (current master fa00183e5f864e54c7fefe9946f1db4124cae3e1) then 

```
$ env
HOME=/home/dÄvis
LANG=lv_LV.UTF-8
LOGNAME=dÄvis
MAIL=/var/spool/mail/dÄvis
USER=dÄvis
$ echo $HOME
/home/dÄvis
```

I guess to fix this with fish's current implementation would need to reread them after locale (`LANG`) change.
